From 15e751e4b79475fb34e4b32a3ca54119b20c564a Mon Sep 17 00:00:00 2001
From: Hongxu Jia <hongxu.jia@windriver.com>
Date: Sat, 17 Aug 2024 21:33:18 +0800
Subject: [PATCH] link libatomic for clang conditionally

Clang emits atomic builtin, explicitly link libatomic conditionally:
- For target build, always link -latomic for clang as usual
- For host build, if host and target have same bit width, cross compiling
  is enabled, and host toolchain is gcc which does not link -latomic;
  if host and target have different bit width, no cross compiling,
  host build is the same with target build that requires to link
  -latomic;

Upstream-Status: Inappropriate [OE specific]

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>
---
 node.gyp                  | 13 ++++++++++++-
 tools/v8_gypfiles/v8.gyp | 15 ++++++++++++---
 2 files changed, 24 insertions(+), 4 deletions(-)

diff --git a/node.gyp b/node.gyp
--- a/node.gyp
+++ b/node.gyp
@@ -532,7 +532,18 @@
         ],
       }],
       ['OS=="linux" and clang==1', {
-        'libraries': ['-latomic'],
+        'target_conditions': [
+           ['_toolset=="host"', {
+             'conditions': [
+               ['"<!(echo $HOST_AND_TARGET_SAME_WIDTH)"=="0"', {
+                 'libraries': ['-latomic'],
+               }],
+             ],
+           }],
+           ['_toolset=="target"', {
+             'libraries': ['-latomic'],
+           }],
+          ],
       }],
     ],
   },
diff --git a/tools/v8_gypfiles/v8.gyp b/tools/v8_gypfiles/v8.gyp
--- a/tools/v8_gypfiles/v8.gyp
+++ b/tools/v8_gypfiles/v8.gyp
@@ -1325,9 +1325,18 @@
         # to implement atomic memory access.
         # Clang needs it for some atomic operations (https://clang.llvm.org/docs/Toolchain.html#atomics-library).
         ['(OS=="linux" and clang==1) or (v8_current_cpu in ["mips64", "mips64el", "arm", "riscv64", "loong64"])', {
-          'link_settings': {
-            'libraries': ['-latomic', ],
-          },
+          'target_conditions': [
+             ['_toolset=="host"', {
+               'conditions': [
+                 ['"<!(echo $HOST_AND_TARGET_SAME_WIDTH)"=="0"', {
+                   'libraries': ['-latomic'],
+                 }],
+               ],
+             }],
+             ['_toolset=="target"', {
+               'libraries': ['-latomic', ],
+             }],
+           ],
         }],
       ],
     },  # v8_base_without_compiler
